File: src/main/java/no/novari/flyt/history/repository/filters/IntegrationStatisticsQueryFilter.java
package no.novari.flyt.history.repository.filters;

import lombok.Builder;
import org.springframework.kafka.support.JavaUtils;

import java.util.Collection;
import java.util.Optional;
import java.util.StringJoiner;

@Builder
public class IntegrationStatisticsQueryFilter {
    private final Collection<Long> sourceApplicationIds;
    private final Collection<String> sourceApplicationIntegrationIds;
    private final Collection<Long> integrationIds;

    public Optional<Collection<Long>> getSourceApplicationIds() {
        return Optional.ofNullable(sourceApplicationIds);
    }

    public Optional<Collection<String>> getSourceApplicationIntegrationIds() {
        return Optional.ofNullable(sourceApplicationIntegrationIds);
    }

    public Optional<Collection<Long>> getIntegrationIds() {
        return Optional.ofNullable(integrationIds);
    }

    @Override
    public String toString() {
        StringJoiner joiner = new StringJoiner(",", "(", ")");
        JavaUtils.INSTANCE.acceptIfNotNull(
                sourceApplicationIds, sourceApplicationIds -> joiner.add("sourceApplicationIds=" + sourceApplicationIds)
        );
        JavaUtils.INSTANCE.acceptIfNotNull(
                sourceApplicationIntegrationIds, sourceApplicationIntegrationIds -> joiner.add("sourceApplicationIntegrationIds=" + sourceApplicationIntegrationIds)
        );
        JavaUtils.INSTANCE.acceptIfNotNull(
                integrationIds, integrationIds -> joiner.add("integrationIds=" + integrationIds)
        );
        return joiner.toString();
    }

}

File: src/main/kotlin/no/novari/flyt/history/repository/filters/IntegrationStatisticsQueryFilter.kt
package no.novari.flyt.history.repository.filters

data class IntegrationStatisticsQueryFilter(
    val sourceApplicationIds: Collection<Long>? = null,
    val sourceApplicationIntegrationIds: Collection<String>? = null,
    val integrationIds: Collection<Long>? = null,
) {
    companion object {
        @JvmStatic
        fun builder() = Builder()
    }

    class Builder {
        private var sourceApplicationIdsValue: Collection<Long>? = null
        private var sourceApplicationIntegrationIdsValue: Collection<String>? = null
        private var integrationIdsValue: Collection<Long>? = null

        fun sourceApplicationIds(sourceApplicationIds: Collection<Long>?) =
            apply {
                this.sourceApplicationIdsValue = sourceApplicationIds
            }

        fun sourceApplicationIntegrationIds(sourceApplicationIntegrationIds: Collection<String>?) =
            apply {
                this.sourceApplicationIntegrationIdsValue = sourceApplicationIntegrationIds
            }

        fun integrationIds(integrationIds: Collection<Long>?) =
            apply {
                this.integrationIdsValue = integrationIds
            }

        fun build() =
            IntegrationStatisticsQueryFilter(
                sourceApplicationIds = sourceApplicationIdsValue,
                sourceApplicationIntegrationIds = sourceApplicationIntegrationIdsValue,
                integrationIds = integrationIdsValue,
            )
    }
}
